Add search index sort multi field

geichelberger opened the pull request:
The default sorting of the search index could be more intuitive. For example, lowercase letters follow uppercase letters, which leads to funny results that by sorting ascending, the letter `a` comes after `Z`.

By introducing a multi-field `sort` for selected fields, it is possible to ~normalize the string to lowercase, and with asciifolding~ use icu_collation_keyword field type which defaults to DUCET Collation a best effort attempt to better sort results.

https://www.elastic.co/guide/en/elasticsearch/plugins/current/analysis-icu-collation-keyword-field.html

An even better approach would be to use the `icu_collation` filter, which can be configured to use a collation table for a specific language.

Expected:
![grafik](https://github.com/opencast/opencast/assets/35195803/f415c9e4-7020-4103-b4b1-22ba68dd5645)

Current:
![grafik](https://github.com/opencast/opencast/assets/35195803/88b059a7-efbc-485b-9623-9e46d1644162)


* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

geichelberger committed:
Add search index sort multi field

The default sorting of the search index could be more intuitive. For
example, lowercase letters follow uppercase letters, which leads to
funny results that by sorting ascending, the letter `a` comes after `Z`.

By introducing a multi-field `sort` for selected fields, it is
possible to normalize the string to lowercase, and with asciifolding,
better results can be achieved.

An even better approach would be to use `icu_collation` filter, which
can be configured to use a collation table for a specific language.

Signed-off-by: Gregor Eichelberger <gregor.eichelberger@tuwien.ac.at>

ziegenberg commented:
I would suggest going for the `icu_collation` filter which can be configured.

gregorydlogan commented:
Considering that reindexing is already on the table for 16.0, is this going to be moved forward?

geichelberger commented:
It's on my todo list for OC 16

ziegenberg commented:
Installing the ICU plugin is not particularly difficult: `bin/opensearch-plugin install analysis-icu` :)

geichelberger committed:
Replace sort_normalizer with icu_collation_keyword

geichelberger committed:
Add documentation

geichelberger commented:
### Current vs ICU DUCET collation
![grafik](https://github.com/opencast/opencast/assets/35195803/d09da9ab-f685-4cce-95cd-44e92de50377)

geichelberger committed:
Fix search query get order

The getOrder method must check if a sort field variant of the field
is used for ordering.

geichelberger committed:
Refactor sort field integration

This replaces the explicit method for sorting fields via multi-fields
and integrates the logic into the standard method using a lookup table.

ziegenberg commented on docs/guides/admin/docs/installation/debs.md:
```diff
@@ -77,6 +77,10 @@ accomplished by running the following:
 If you wish to use the upstream OpenSearch repository directly be aware that Opencast only supported with OpenSearch 1.x
 and will not work with OpenSearch 2.x yet.  Future support for this is forthcoming.
 
+To install the ICU plugin for OpenSearch, run the following:
```
We should tell the people why we need the ICU Plugin. Currently, this dependency is not justified in the docs.

gregorydlogan replied:
This should also clarify that things *will not work* without this call. I would absolutely skip this as a naive just-testing adopter, and then be confused when it all looks like it should work, but doesn't.

KatrinIhler commented:
The general tendency in the dev meeting was that making people re-index the whole thing instead of just the search part would be acceptable, especially since OC 16 is already a release with many breaking changes. So we probably don't want this to be configurable or only do the search index, but go all in.

gregorydlogan commented on docs/guides/admin/docs/installation/rpm-el.md:
````diff
@@ -70,6 +70,8 @@ Make sure to install it on the node which also serves the admin interface.
 dnf install elasticsearch-oss
 ```
 
+
````
Presumably the rpms need the same call made as the debs?

gregorydlogan reviewed (commented):
Reminder that we need to incorporate this into #5597, or to incorporate that mapping file into this one, depending on who gets merged first.

geichelberger committed:
Extend documentation

geichelberger commented:
It would be easy to update the index without a complete rebuild.

**Series Index Example:**

`PUT /opencast_series/_mapping`
``` json
{
  "properties": {
    "title": {
      "type": "keyword",
      "fields": {
        "sort": {
          "type": "icu_collation_keyword",
          "index": false
        }
      }
    },
    "description": {
      "type": "keyword",
      "fields": {
        "sort": {
          "type": "icu_collation_keyword",
          "index": false
        }
      }
    },
    "subject": {
      "type": "keyword",
      "fields": {
        "sort": {
          "type": "icu_collation_keyword",
          "index": false
        }
      }
    },
    "creator": {
      "type": "keyword",
      "fields": {
        "sort": {
          "type": "icu_collation_keyword",
          "index": false
        }
      }
    },
    "organizers": {
      "copy_to": "organizer",
      "type": "keyword",
      "fields": {
        "sort": {
          "type": "icu_collation_keyword",
          "index": false
        }
      }
    },
    "contributors": {
      "copy_to": "contributor",
      "type": "keyword",
      "fields": {
        "sort": {
          "type": "icu_collation_keyword",
          "index": false
        }
      }
    },
    "publisher": {
      "type": "keyword",
      "fields": {
        "sort": {
          "type": "icu_collation_keyword",
          "index": false
        }
      }
    },
    "rights_holder": {
      "type": "keyword",
      "fields": {
        "sort": {
          "type": "icu_collation_keyword",
          "index": false
        }
      }
    }
  }
}
```

Followed by:

` POST /opencast_series/_update_by_query?refresh&conflicts=proceed`

Update by query has many possible options. The version above is the simplest, and it may be enough, even if you have 200k entries, which are peanuts for elasticsearch.

gregorydlogan commented:
This seems like something we should add to the docs as an option for those leery of doing a full rebuild!

gregorydlogan commented:
@geichelberger this needs a note about the reindex, but then it should go into `develop` so it can be in for 16.0

geichelberger committed:
Add release note and reindex alternative

mtneug commented:
Is this still targeting 16?

gregorydlogan commented on docs/upgrade/15_to_16/update_event_mapping.json:
```diff
@@ -0,0 +1,83 @@
+{
```
These should be in 16_to_17 right?

geichelberger replied:
Yes, I moved the files.

gregorydlogan reviewed (approved):
Works as expected.  Should probably move the upgrade scripts, right?

geichelberger committed:
Move upgrade script

KatrinIhler commented:
FYI: Branch cut for OC 17 is on November 6!
